omnigan/trainer.py

```diff
@@ -1773,3 +1292,280 @@ def eval_images(self, mode, domain):
                 )
 
         return 0
+
+    def masker_c_loss(self, z, target, for_="G"):
+        assert for_ in {"G", "D"}
+        full_loss = 0
+        # -------------------
+        # -----  Depth  -----
+        # -------------------
+        one_hot = self.opts.classifier.loss != "cross_entropy"
+        output_classifier = self.C(z)
+        # Cross entropy loss (with sigmoid) with fake labels to fool C
+        loss = self.losses["G"]["classifier"](
+            output_classifier, fake_domains_to_class_tensor(target, one_hot),
+        )
+        loss *= self.opts.train.lambdas.G.classifier
+        full_loss += loss
+
+        return full_loss
+
+    def assert_z_matches_x(self, x, z):
+        assert x.shape[0] == (
+            z.shape[0] if not isinstance(z, (list, tuple)) else z[0].shape[0]
+        ), "x-> {}, z->{}".format(
+            x.shape, z.shape if not isinstance(z, (list, tuple)) else z[0].shape
+        )
+
+    def masker_d_loss(self, x, z, target, for_="G"):
+        assert for_ in {"G", "D"}
+        self.assert_z_matches_x(x, z)
+        assert x.shape[0] == target.shape[0]
+        full_loss = 0
+        # -------------------
+        # -----  Depth  -----
+        # -------------------
+
+        prediction = self.G.decoders["d"](z)
+        loss = self.losses["G"]["tasks"]["d"](prediction, target)
+        loss *= self.opts.train.lambdas.G.d.main
+
+        full_loss += loss
+
+        return full_loss
+
+    def masker_s_loss(self, x, z, target, domain, for_="G"):
+        assert for_ in {"G", "D"}
+        assert domain in {"r", "s"}
+        self.assert_z_matches_x(x, z)
+        assert x.shape[0] == target.shape[0] if target is not None else True
+        full_loss = 0
+        softmax_preds = None
+        # --------------------------
+        # -----  Segmentation  -----
+        # --------------------------
+        pred = None
+        if for_ == "G" or self.opts.gen.s.use_advent:
```

Comment:
I'm not sure of the whole disposition of the function. Wouldn't it be clearer to have a big "if for_=="G"; else"? (For example, I think lines from 1348 to 1375 could all be together in a 'if for_ == "G"').

Reply:
Done